cmd/admin-handlers.go

```diff
@@ -213,3 +219,156 @@ func (adminAPI adminAPIHandlers) ClearLocksHandler(w http.ResponseWriter, r *htt
 	// Reply with list of locks cleared, as json.
 	writeSuccessResponseJSON(w, jsonBytes)
 }
+
+func validateHealQueryParams(vars url.Values) (string, string, string, string, int, APIErrorCode) {
+	bucket := vars.Get(string(mgmtBucket))
+	prefix := vars.Get(string(mgmtPrefix))
+	marker := vars.Get(string(mgmtMarker))
+	delimiter := vars.Get(string(mgmtDelimiter))
+	maxKeyStr := vars.Get(string(mgmtMaxKey))
+
+	// N B empty bucket name is invalid
+	if !IsValidBucketName(bucket) {
+		return "", "", "", "", 0, ErrInvalidBucketName
+	}
+
+	// empty prefix is valid.
+	if !IsValidObjectPrefix(prefix) {
+		return "", "", "", "", 0, ErrInvalidObjectName
+	}
+
+	// check if maxKey is a valid integer.
+	maxKey, err := strconv.Atoi(maxKeyStr)
+	if err != nil {
+		return "", "", "", "", 0, ErrInvalidMaxKeys
+	}
+
+	// Validate prefix, marker, delimiter and maxKey.
+	apiErr := validateListObjectsArgs(prefix, marker, delimiter, maxKey)
+	if apiErr != ErrNone {
+		return "", "", "", "", 0, apiErr
+	}
+
+	return bucket, prefix, marker, delimiter, maxKey, ErrNone
+}
+
+// ListObjectsHealHandler - GET /?heal&bucket=mybucket&prefix=myprefix&marker=mymarker&delimiter=&mydelimiter&maxKey=1000
+// - bucket is mandatory query parameter
+// - rest are optional query parameters
+// List upto maxKey objects that need healing in a given bucket matching the given prefix.
+func (adminAPI adminAPIHandlers) ListObjectsHealHandler(w http.ResponseWriter, r *http.Request) {
+	// Get object layer instance.
+	objLayer := newObjectLayerFn()
+	if objLayer == nil {
+		writeErrorResponse(w, ErrServerNotInitialized, r.URL)
+		return
+	}
+
+	// Validate request signature.
+	adminAPIErr := checkRequestAuthType(r, "", "", "")
+	if adminAPIErr != ErrNone {
+		writeErrorResponse(w, adminAPIErr, r.URL)
+		return
+	}
+
+	// Validate query params.
+	vars := r.URL.Query()
+	bucket, prefix, marker, delimiter, maxKey, adminAPIErr := validateHealQueryParams(vars)
+	if adminAPIErr != ErrNone {
+		writeErrorResponse(w, adminAPIErr, r.URL)
+		return
+	}
+
+	// Get the list objects to be healed.
+	objectInfos, err := objLayer.ListObjectsHeal(bucket, prefix, marker, delimiter, maxKey)
+	if err != nil {
+		writeErrorResponse(w, toAPIErrorCode(err), r.URL)
+		return
+	}
+
+	listResponse := generateListObjectsV1Response(bucket, prefix, marker, delimiter, maxKey, objectInfos)
```

Comment:
You should use a separate function for this for clarity.  We shouldn't reply on the same XML, we need little more data than what is returned right now. For example objects that needs to be healed should return information like 

- How many shards are available? 
- Can healing be performed or not - like object is lost and cannot be recovered?